[#43697] add storage files breadcrumbs

Kharonus opened the pull request:
- https://community.openproject.org/work_packages/43697
- added spot breadcrumbs styles
- added storage files breadcrumb component

Kharonus committed:
[#43697] add storage files breadcrumbs

- https://community.openproject.org/work_packages/43697
- added spot breadcrumbs styles
- added storage files breadcrumb component

Kharonus committed:
[#43697] added breadcrumb collapsing logic

Kharonus committed:
[#43697] moved breadcrumbs to general shared component

Kharonus committed:
[#43697] added mobile style

Kharonus committed:
[#43697] amended mobile switches

apfohl commented on frontend/src/app/shared/components/breadcrumbs/breadcrumbs.ts:
```diff
@@ -0,0 +1,39 @@
+  navigate?:() => void;
+}
+
+export class Breadcrumbs {
```
To support the pattern, I suggest `BreadcrumbsViewModel`. Otherwise people could get confused, why you just wrap an array in a class.

Kharonus replied:
TBH I'm not a big fan of adding a suffix to just explain, what it is used for. this is too much of an Hungarian Notation for me. ^^
Maybe we should clarify before, if a view model is something that is used, or defined. Or, with other words, would I be allowed to use the same model in another context, where it is not used as a view model.

I wrote this model with this thought: `Breadcrumbs` is a model, it consists of an array of single crumbs. We may also add more to that model. I just didn't need it until now (like configurational parameters, collapsing depth, etc.)

apfohl replied:
But the we should also call the input parameter of the component something else as well.

Kharonus replied:
changed naming to `content`.

apfohl commented on frontend/src/app/spot/styles/sass/components/breadcrumbs.sass:
```diff
@@ -0,0 +1,35 @@
+  display: flex
+  align-items: baseline
+
+  &--crumb
+    display: flex
+    align-items: center
+    flex-shrink: 0
+    min-width: 0
+
+    > .spot-icon
+      width: $spot-spacing-1-5
+      height: $spot-spacing-1-5
+
+    &_shrink
+      flex-shrink: 1
+
+    &-action-item
+      flex-shrink: 1
+      white-space: nowrap
+      overflow: hidden
+      text-overflow: ellipsis
+
+      &:not(:first-child)
+        margin-left: $spot-spacing-0-25
+
+
+    &-divider
+      font-size: 0.5rem
+      flex-shrink: 0
+
+    &:not(:first-child)
+      margin-left: $spot-spacing-0-25
```
I'm confused about the BEM styles again. What are the naming rules we follow?

Kharonus replied:
What is it, you are confused about? Can you pinpoint it?

apfohl replied:
single vs. double underscore. And this website: https://en.bem.info/methodology/naming-convention/

Kharonus replied:
it is still a bit confusing, but we use the following rules:

- block is kebab case
- block is separated from element with `--`
- modifier are separated with single `_`
- in variable names we use `-` for a space (separator) and a `_` as a replacement for the decimal dot in numbers

Kharonus committed:
[#43697] fixed PR issues

- added storage icon factory
- renaming of view model classes and variables
- fixes of spot variable usage

Kharonus committed:
[#43697] small cleanups

b12f commented on frontend/src/app/shared/components/breadcrumbs/breadcrumbs.component.ts:
```diff
@@ -0,0 +1,47 @@
+import { BreadcrumbsContent } from 'core-app/shared/components/breadcrumbs/breadcrumbs-content';
+
+@Component({
+  selector: 'op-breadcrumbs',
```
If this is a spot component, then this selector should be `spot-breadcrumb` and the files should be in `/frontend/src/app/spot/components` (including stories ;))

Kharonus replied:
true, will move it and add stories.

Kharonus replied:
moved it. But I just spoke with @psatyal and I think, we should write the stories in a different branch. I nevertheless start writing them tomorrow, but I rather wouldn't want to delay this PR with the writing of the stories. Is that okay for you, @b12f ?

b12f replied:
yup all good

Kharonus committed:
[#43697] moved op-breadcrumbs to spot-breadcrumbs

b12f commented on frontend/src/app/spot/components/breadcrumbs/breadcrumbs-content.ts:
```diff
@@ -0,0 +1,39 @@
+  navigate?:() => void;
+}
+
+export class BreadcrumbsContent {
```
Please do this as an interface instead of a class. Using classes just makes it harder to serialize/deserialize from JSON.

Kharonus replied:
but how do I make an interface immutable? What is the effect you mention, that it is harder to serialize? does it take more time?
This class is used as a view model, meaning it contains any information, the breadcrumbs component needs to render. So, its usage usually should not have any serialization involved, and the immutability is a nice thing to have.

b12f replied:
You could freeze the object, but I don't think there's a reason to make this immutable. 99% of cases we'll just be handing over an array from parent to child

b12f replied:
> What is the effect you mention, that it is harder to serialize?

You can't run `JSON.stringify` on this data. It might feel weird to think about this, but this is stateful data and we try to keep all stateful data serializable. I understand it's not possible for the callbacks in this interface (but they're not even used anyway, so you might as well delete them instead of offering a "fake" API.)

Kharonus replied:
immutability is key for view models, as it prevents the parent from altering the data in it while it is already "in use" for the rendered component. this makes the view more stable and IMHO easier to use, as you have only one binding and to alter it, you have to create a new view model. This way, you can always predict, how the view will look afterwards, there is no "hidden" data, which was bound to the view in some other initialization within the parent.

the callbacks are in use now, the template's buttons have the click handler defined.

And why is it, you should not be able to run `JSON.stringify` on a js object created with a class constructor?
![image](https://user-images.githubusercontent.com/38206611/197725496-eb404306-f548-49f6-b81b-a343c7251ed7.png)

Kharonus replied:
discussion delayed to include other devs ;)

b12f commented on frontend/src/app/spot/components/breadcrumbs/breadcrumbs.component.ts:
```diff
@@ -0,0 +1,47 @@
+
+@Component({
+  selector: 'spot-breadcrumbs',
+  templateUrl: './breadcrumbs.html',
```
```suggestion
  templateUrl: './breadcrumbs.component.html',
```

b12f commented on frontend/src/app/spot/components/breadcrumbs/breadcrumbs.html:
```diff
@@ -0,0 +1,72 @@
+    <button
+      type="button"
+      class="spot-link spot-breadcrumbs--crumb-action-item"
+      [title]="crumb.text"
```
This should have some kijnd of click listener, no?

Kharonus replied:
true, currently there is no functionality attached yet, but the click handler can get added.

b12f replied:
I'd keep it a normal div until then so that it can't get focus

b12f commented on frontend/src/app/spot/components/breadcrumbs/breadcrumbs.html:
```diff
@@ -0,0 +1,72 @@
+
+  <div
+    *ngIf="i === 1 && content.crumbs.length > 4"
+    class="spot-breadcrumbs--crumb hidden-for-mobile"
```
Please write spot-specific rules instead of using the `hidden-for-` utility classes.

b12f replied:
In fact, I don´t see a mobile/desktop specific version in figma? But I might just be missing it. In any case this template feels like it is way too big for what the component does, though maybe that's warranted.

b12f replied:
e.g. you can just render one list instead of two and hide specific elements based on classes (e.g. add `spot-breadcrumb--crumb_first` or something, and select based on those)

Kharonus replied:
Here you can find the mobile version: https://www.figma.com/file/gtLQfPe09X7XugAH8L7dTy/Nextcloud-Integration?node-id=2541%3A69682

Kharonus replied:
I have the same feeling like you, but...

When I tried to put everything into one list and hide specific elements, the switches between mobile and desktop went crazy, due to the fact, that mobile and desktop just behave soooo different. It is not only display logic, but also shrinkability and content.

In the end I found the `ng-container` with the two sections (desktop items and mobile items) by far the most readable solution. But if you want to pair to find a better solution, please let me know.

Kharonus replied:
reworked the template after discussing the possibilities to switch display logic in SASS classes.

b12f commented on frontend/src/app/shared/components/file-links/storage-icons/storage-icon.factory.ts:
```diff
@@ -0,0 +1,37 @@
+
+import { storageIconMappings } from 'core-app/shared/components/file-links/storage-icons/icon-mappings';
+
+export function getIconForStorageType(storageType?:string):string {
```
Is this a factory? I'd just call this file `get-icon-for-storage-type.ts` and default export this function.

Kharonus replied:
well, it was kind of copied from another factory function, that produced real objects, this one only produces strings, so it might not stand as a "factory".

Kharonus committed:
[#43697] fixed PR issues

Kharonus committed:
[#43697] refactored crumbs template

b12f commented on frontend/src/app/spot/components/breadcrumbs/breadcrumbs.component.html:
```diff
@@ -0,0 +1,44 @@
+    *ngIf="(first && !last) || i === content.crumbs.length - 2 || i === content.crumbs.length - 3"
+    class="spot-breadcrumbs--crumb"
+    [ngClass]="{
+      'spot-breadcrumbs--crumb_first': first,
```
maybe we should call this `_root`

Kharonus replied:
renamed

b12f reviewed (approved):
LGTM to merge this, let's talk about immutability and state with @oliverguenther and @bsatarnejad next week

Kharonus committed:
[#43697] replaced first with root
